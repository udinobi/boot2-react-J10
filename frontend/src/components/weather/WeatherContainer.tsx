import { Col, Row } from "antd"
import * as moment from "moment-timezone"
import React from "react"

import { option, Option } from "ts-option"

import Description from "./Description"
import Humidity from "./Humidity"
import LastUpdate from "./LastUpdate"
import Sunrise from "./Sunrise"
import Sunset from "./Sunset"
import Temperature from "./Temperature"
import TemperatureMax from "./TemperatureMax"
import TemperatureMin from "./TemperatureMin"

import { LargeText } from "../lib/StyledText"

import { Location } from "../../store/location/city/types"

import { WeatherData } from "../../store/weather/types"

export interface Data {
    location: Location
    lastUpdate: Option<Date>
    weatherData: WeatherData
}

export const frmt = (val: number) => ("0" + val).slice(-2)

export const time = (t: number) => {
    const dt = new Date(t)
    return `${frmt(dt.getUTCHours())}:${frmt(dt.getUTCMinutes())}:${frmt(dt.getUTCSeconds())}`
}

export default (props: Data) => {

    const offsetInMinutes = moment.tz(props.location.tz).utcOffset()
    const offsetInMillis = offsetInMinutes * 60 * 1000

    const sunrise = props.weatherData.sys.sunrise * 1000 + offsetInMillis
    const sunset = props.weatherData.sys.sunset * 1000 + offsetInMillis

    const offsetInHours = Math.round(offsetInMinutes / 60)
    const GMT = ` GMT${offsetInHours < 0 ? "" : "+"}${offsetInHours}`

    return <div>
        <Row>
            <LargeText>{props.location.name} {"\u00a0"}({props.location.country.code})</LargeText>
            <LastUpdate lastUpdate={props.lastUpdate} />
        </Row>
        <Row>
            <Col xs={24} sm={24} md={24} lg={15}>
                <Row>
                    <Col>
                        <Description weather={props.weatherData.weather} />
                    </Col>
                </Row>
                <Row>
                    <Col span={12}>
                        <Temperature weatherData={option(props.weatherData)} />
                    </Col>
                    <Col span={12}>
                        <div style={{ float: "right", marginTop: 8 }}>
                            <Humidity humidity={props.weatherData.main.humidity} />
                            <TemperatureMin minTemperature={props.weatherData.main.temp_min} />
                            <TemperatureMax maxTemperature={props.weatherData.main.temp_max} />
                        </div>
                    </Col>
                </Row>
            </Col>
            <Col xs={24} sm={24} md={24} lg={9}>
                <Sunrise sunrise={`${time(sunrise)}${GMT}`} />
                <Sunset sunset={`${time(sunset)}${GMT}`} />
            </Col>
        </Row>
    </div>
}
