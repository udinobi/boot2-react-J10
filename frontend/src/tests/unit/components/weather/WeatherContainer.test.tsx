import { shallow } from "enzyme"
import toJson from "enzyme-to-json"
import React from "react"

import { none } from "ts-option"

import WeatherContainer, { Data } from "../../../../components/weather/WeatherContainer"

describe("--> Snapshot for Components > weather > 'WeatherContainer'", () => {
    let props: Data

    beforeEach(() => {
        props = {
            lastUpdate: none,
            location: {
                coord: { lat: "", lon: "" },
                country: { code : "IT", name: "Italy" },
                geoId: 0,
                name: "",
                tz: ""
            },
            weatherData: {
                coord: { lat: 0, lon: 0 },
                main: {
                    grnd_level: 0,
                    humidity: 0,
                    pressure: 0,
                    sea_level: 0,
                    temp: 0,
                    temp_max: 0,
                    temp_min: 0,
                },
                sys: { sunrise: 0, sunset: 0 },
                weather: [ {
                    description: "",
                    icon: "",
                    id: 0,
                    main: ""
                } ]
            }
        }
    })

    it("matches", () => {
        const wrapper = shallow(<WeatherContainer {...props}/>)
        expect(toJson(wrapper)).toMatchSnapshot()
    })
})
