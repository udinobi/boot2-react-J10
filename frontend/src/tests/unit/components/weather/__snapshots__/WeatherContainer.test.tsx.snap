// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`--> Snapshot for Components > weather > 'WeatherContainer' matches 1`] = `
<div>
  <Row
    gutter={0}
  >
    <Styled(styled.div)>
       
      Â 
      (
      IT
      )
    </Styled(styled.div)>
    <Component
      lastUpdate={None {}}
    />
  </Row>
  <Row
    gutter={0}
  >
    <Col
      lg={15}
      md={24}
      sm={24}
      xs={24}
    >
      <Row
        gutter={0}
      >
        <Col>
          <Component
            weather={
              Array [
                Object {
                  "description": "",
                  "icon": "",
                  "id": 0,
                  "main": "",
                },
              ]
            }
          />
        </Col>
      </Row>
      <Row
        gutter={0}
      >
        <Col
          span={12}
        >
          <Component
            weatherData={
              Some {
                "_value": Object {
                  "coord": Object {
                    "lat": 0,
                    "lon": 0,
                  },
                  "main": Object {
                    "grnd_level": 0,
                    "humidity": 0,
                    "pressure": 0,
                    "sea_level": 0,
                    "temp": 0,
                    "temp_max": 0,
                    "temp_min": 0,
                  },
                  "sys": Object {
                    "sunrise": 0,
                    "sunset": 0,
                  },
                  "weather": Array [
                    Object {
                      "description": "",
                      "icon": "",
                      "id": 0,
                      "main": "",
                    },
                  ],
                },
              }
            }
          />
        </Col>
        <Col
          span={12}
        >
          <div
            style={
              Object {
                "float": "right",
                "marginTop": 8,
              }
            }
          >
            <Component
              humidity={0}
            />
            <Component
              minTemperature={0}
            />
            <Component
              maxTemperature={0}
            />
          </div>
        </Col>
      </Row>
    </Col>
    <Col
      lg={9}
      md={24}
      sm={24}
      xs={24}
    >
      <Component
        sunrise="00:00:00 GMT+0"
      />
      <Component
        sunset="00:00:00 GMT+0"
      />
    </Col>
  </Row>
</div>
`;
